File: src/main/java_old/net/orandja/vanillawaifu/mods/ProtectBlock.java
package net.orandja.vanillawaifu.mods;

import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.orandja.vanillawaifu.VanillaWaifu;

/**
 * Related Mixins:
 * net.orandja.vanillawaifu.mixin.ExplosionMixin
 * net.orandja.vanillawaifu.mixin.HopperBlockEntityMixin
 * net.orandja.vanillawaifu.mixin.PlayerEntityMixin
 * */
public class ProtectBlock {

    public interface ProtectDestruction {
        boolean canDestroy(World world, BlockPos pos, Object entity);
    }

    public interface ProtectExtract {
        boolean preventExtract(World world, BlockPos pos);
    }

    public static boolean canDestroy(World world, BlockPos pos, Object entity) {
        for (ProtectDestruction protector : VanillaWaifu.PROTECTORS) {
            if(protector.canDestroy(world, pos, entity)) {
                return true;
            }
        }
        return false;
    }

    public static boolean preventExtract(World world, BlockPos pos) {
        for (ProtectExtract protector : VanillaWaifu.HOPPER_EXTRACT) {
            if(protector.preventExtract(world, pos)) {
                return true;
            }
        }
        return false;
    }

    public static boolean preventExtract(World world, double x, double y, double z) {
        return preventExtract(world, new BlockPos(x, y, z));
    }
}

File: src/main/kotlin/net/orandja/vw/mods/ProtectBlock.kt
package net.orandja.vw.mods

import net.minecraft.entity.Entity
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class ProtectBlock {

    companion object {

        // returns true if the player can destroy the block
        val DESTROY: ArrayList<(world: World, pos: BlockPos, entity: Entity) -> Boolean> = ArrayList()
        // returns true if the explosion can destroy the block
        val EXPLOSIONS: ArrayList<(world: World, pos: BlockPos) -> Boolean> = ArrayList()
        // returns false if the hopper can extract item from block **Might be changed in the future
        val EXTRACTORS: ArrayList<(world: World, pos: BlockPos) -> Boolean> = ArrayList()

        @JvmStatic
        fun canDestroy(world: World, pos: BlockPos, player: Any): Boolean {
            return DESTROY.find { protector -> protector.invoke(world, pos, player as Entity) } != null
        }

        @JvmStatic
        fun canExplode(world: World, pos: BlockPos): Boolean {
            return EXPLOSIONS.find { protector -> protector.invoke(world, pos) } != null
        }

        fun preventsExtract(world: World, pos: BlockPos): Boolean {
            return EXTRACTORS.find { extractor -> extractor.invoke(world, pos) } != null
        }

        @JvmStatic
        fun preventsExtract(world: World, x: Double, y: Double, z: Double): Boolean {
            return preventsExtract(world, BlockPos(x, y, z))
        }
    }
}